corehq/apps/userreports/tests/test_choice_provider.py

```diff
@@ -537,18 +537,28 @@ def make_web_user_with_registry_role(cls, email, domain, has_registry_access=Fal
 
     @classmethod
     def setUpClass(cls):
-        super(DomainChoiceProviderTest, cls).setUpClass()
+        super().setUpClass()
         cls.domain_a = create_domain(name="A")
```

Comment:
These refactors remind me of '[Resource Allocation is Initialization](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization)' (RAII) from C++, which is a complicated term essentially meaning that the constructor is in charge of acquiring resources, and the destructor automatically frees those resources. The upside of this is that, in normal usage, you can now no longer have any resource leaks, as it's all managed automatically.

Prior to this concept, resources were managed much like what's being written here:
```python
# acquire resource
# release resource
```

The downside is that this isn't as automatic and is error-prone. To avoid duplication here, we've created a loop to do this, which is even more error-prone (if a new domain is added, but not added to the deletion list). And we see some duplication with the user deletion logic later on.

I'd consider adopting the RAII concept here, instead. Instead of doing:
```python
cls.domain = create_domain(...)
cls.addClassCleanup(cls.domain.delete)
```

I'd recommend moving these creation methods into their own self-managed class functions, i.e.:
```python
@classmethod
create_domain(cls, ...):
  domain = create_domain(...)
  cls.addClassCleanup(domain.delete)

@classmethod
def setUpClass(cls):
  cls.domain = cls.create_domain(...)
  # no danger now of leaking the domain
```

Reply:
So before this PR, setup code created objects, and tear down code deleted those created objects. With the current implementation, setup code creates objects and specifies that they should be cleaned up. And you are advocating for setup code calling helper methods that both create and specify cleanup of those objects to remove this burden from the test itself.

I agree that is the optimal solution, but don't see that as a blocker to this improvement. Based on your approval, I think you feel similarly, but let me know if that isn't the case. The only reason I'd want to merge without making that change is because this blocks other work to improve our domain deletion process.

Reply:
Yeah, I'm not going to hold up approval based on that issue, just wanted to point it out as a potential option